File: python/group.py
# coding: utf-8
# group.py
# 同じFillTypeのRecordの集まり

from __future__ import annotations
from typing import Optional, Iterator, List, Tuple
from itertools import *

from VCFImpFamilyRecord import FillType, VCFImpFamilyRecord
from VCFFillableRecord import VCFFillableRecord
from RecordSet import RecordSet, RecordSetSmall

Group = Tuple[FillType, List[VCFFillableRecord]]


#################### Groups ####################

class Groups:
	def __init__(self, groups: list[Group]):
		self.groups: list[Group] = groups
	
	def __len__(self) -> int:
		return len(self.groups)
	
	def find_prev_record(self,
						i: int, g: FillType) -> Optional[VCFFillableRecord]:
		key, records = self.groups[i]
		for j in range(i-1, -1, -1):
			key, records = self.groups[j]
			if key == g:
				return records[-1]
		else:
			return None
	
	def find_next_record(self,
						i: int, g: FillType) -> Optional[VCFFillableRecord]:
		key, records = self.groups[i]
		for j in range(i+1, len(self)):
			key, records = self.groups[j]
			if key == g:
				return records[0]
		else:
			return None
	
	def generate_record_sets(self) -> Iterator[RecordSet]:
		for i, (key, records) in enumerate(self.groups):
			if key == FillType.MAT or key == FillType.PAT:
				continue
			prev_mat_record = self.find_prev_record(i, FillType.MAT)
			next_mat_record = self.find_next_record(i, FillType.MAT)
			prev_pat_record = self.find_prev_record(i, FillType.PAT)
			next_pat_record = self.find_next_record(i, FillType.PAT)
			for record in records:
				yield RecordSet(record, prev_mat_record,
							next_mat_record, prev_pat_record, next_pat_record)
	
	def generate_record_sets_small(self) -> Iterator[RecordSet]:
		for i, (key, records) in enumerate(self.groups):
			if key == FillType.MAT or key == FillType.PAT:
				continue
			prev_mat_record = self.find_prev_record(i, FillType.MAT)
			next_mat_record = self.find_next_record(i, FillType.MAT)
			prev_pat_record = self.find_prev_record(i, FillType.PAT)
			next_pat_record = self.find_next_record(i, FillType.PAT)
			for record in records:
				yield RecordSetSmall(record, prev_mat_record,
							next_mat_record, prev_pat_record, next_pat_record)
	
	@staticmethod
	def create(records: list[VCFFillableRecord]) -> Groups:
		groups = []
		for g, v in groupby(records, key=lambda r: r.type):
			groups.append((g, list(v)))
		return Groups(groups)

File: src/group.cpp
// group.cpp
#include "../include/group.h"
#include "../include/VCFFillableRecord.h"

using namespace std;


//////////////////// Groups ////////////////////

VCFFillableRecord *Groups::find_prev_record(size_t i, FillType g) const {
	if(i == 0)
		return NULL;
	
	for(size_t j = i-1; j < i; --j) {	// underflow
		const FillType	key = get_group(j).first;
		const auto&	records = get_group(j).second;
		if(key == g)
			return records.back();
	}
	return NULL;
}

VCFFillableRecord *Groups::find_next_record(size_t i, FillType g) const {
	for(size_t j = i+1; j < size(); ++j) {
		const FillType	key = get_group(j).first;
		const auto&	records = get_group(j).second;
		if(key == g)
			return records.front();
	}
	return NULL;
}

vector<RecordSet> Groups::create_record_sets() const {
	vector<RecordSet>	record_sets;
	for(size_t i = 0; i < size(); ++i) {
		const FillType	key = get_group(i).first;
		const auto&	records = get_group(i).second;
		if(key == FillType::MAT || key == FillType::PAT)
			continue;
		auto	*prev_mat_record = find_prev_record(i, FillType::MAT);
		auto	*next_mat_record = find_next_record(i, FillType::MAT);
		auto	*prev_pat_record = find_prev_record(i, FillType::PAT);
		auto	*next_pat_record = find_next_record(i, FillType::PAT);
		for(auto p = records.begin(); p != records.end(); ++p) {
			record_sets.push_back(RecordSet(*p, prev_mat_record,
						next_mat_record, prev_pat_record, next_pat_record));
		}
	}
	return record_sets;
}

Groups *Groups::create(const vector<VCFFillableRecord *>& records) {
	vector<Group>	groups;
	auto	current_type = records.front()->get_type();
	vector<VCFFillableRecord *>	group(1U, records.front());
	for(auto p = records.begin() + 1; p != records.end(); ++p) {
		auto	*record = *p;
		if(record->get_type() != current_type) {
			groups.push_back(Group(current_type, group));
			group.clear();
			current_type = record->get_type();
		}
		group.push_back(record);
	}
	groups.push_back(Group(current_type, group));
	return new Groups(groups);
}
